server/src/main/java/org/apache/iotdb/db/metadata/mtree/store/disk/schemafile/SchemaFile.java

```diff
@@ -277,157 +267,213 @@ public void writeMNode(IMNode node) throws MetadataException, IOException {
-
-          // delete and modify original infos
-          curPage.deleteSegment(curSegIdx);
-          setNodeAddress(node, newSegAddr);
-          updateParentalRecord(node.getParent(), node.getName(), newSegAddr);
-          dirtyPages.putIfAbsent(curPage.getPageIndex(), curPage);
-        } else {
-          // already full page segment, write updated record to another applicable segment or a
-          // blank new one
-          long existedSegAddr =
-              getApplicableLinkedSegments(curPage, curSegIdx, entry.getKey(), childBuffer);
-          if (existedSegAddr < 0) {
-            // get a blank new page and add it to linked segment list right behind curPage
-            ISchemaPage newPage = getMinApplicablePageInMem(newSegSize);
-            existedSegAddr =
-                getGlobalIndex(newPage.getPageIndex(), newPage.allocNewSegment(newSegSize));
-
-            long nextSegAddr = curPage.getNextSegAddress(curSegIdx);
-            if (nextSegAddr != -1) {
-              ISchemaPage nextPage = getPageInstance(getPageIndex(nextSegAddr));
-              nextPage.setPrevSegAddress(getSegIndex(nextSegAddr), existedSegAddr);
-              dirtyPages.putIfAbsent(nextPage.getPageIndex(), nextPage);
-            }
-
-            newPage.setNextSegAddress(getSegIndex(existedSegAddr), nextSegAddr);
-            newPage.setPrevSegAddress(getSegIndex(existedSegAddr), actualSegAddr);
-
-            curPage.setNextSegAddress(getSegIndex(actualSegAddr), existedSegAddr);
+        if (curPage.getAsSegmentedPage().getSegmentSize(getSegIndex(actualAddress))
+            == SchemaPage.SEG_MAX_SIZ) {
+          // split and update higer nodes
+          ISchemaPage splPage = getMinApplSegmentedPageInMem(SchemaPage.SEG_MAX_SIZ);
+          splPage.getAsSegmentedPage().allocNewSegment(SchemaPage.SEG_MAX_SIZ);
+          sk = curPage.getAsSegmentedPage().splitWrappedSegment(null, null, splPage, false);
+          curPage
+              .getAsSegmentedPage()
+              .setNextSegAddress((short) 0, getGlobalIndex(splPage.getPageIndex(), (short) 0));
+
+          // update on page where it exists
+          if (entry.getKey().compareTo(sk) >= 0) {
+            splPage.getAsSegmentedPage().update((short) 0, entry.getKey(), childBuffer);
+          } else {
+            curPage.getAsSegmentedPage().update((short) 0, entry.getKey(), childBuffer);
           }
 
-          ISchemaPage existedPage = getPageInstance(getPageIndex(existedSegAddr));
-          existedPage.write(getSegIndex(existedSegAddr), entry.getKey(), childBuffer);
-          curPage.removeRecord(getSegIndex(actualSegAddr), entry.getKey());
-          dirtyPages.putIfAbsent(curPage.getPageIndex(), curPage);
-          dirtyPages.putIfAbsent(existedPage.getPageIndex(), existedPage);
+          // insert index entry upward
+          if (treeTrace.get()[0] < 1) {
+            ISchemaPage trsPage = getMinApplSegmentedPageInMem(SchemaPage.SEG_MAX_SIZ);
+            trsPage
```

Comment:
It seems the ````treeTrace``` is used for schemaFile write operation. Since there's only one thread doing writing when executing MTree flush, the ```ThreadLocal``` seems unnecessary. Or we can simply use a stack to implement the function.

Reply:
As mentioned above, ```ThreadLocal``` is no longer contained.